Alternate SNMP plugin

phemmer opened the pull request:
This is a proposal for an alternate SNMP plugin.

I wrote this because of issues like #1371, #808 & #1361
I also found the configuration of the existing SNMP plugin to be overly complex & confusing.

This new plugin has:
* A vastly simpler config
* Supports SNMP tables properly (#1371 & #808)
* Supports converting strings to numeric types (#1361 & #1151)
* Supports adding tags using SNMP values (#997)
* Supports SNMP v1 (#1163) & v3 (#856)
* Does not require an external file (`snmptranslate_file`)
* Should also fix #1043 & #961 

### Example usage:

SNMP data:
```
.1.2.3.0.0.1.0 octet_str "foo"
.1.2.3.0.0.1.1 octet_str "bar"
.1.2.3.0.0.102 octet_str "bad"
.1.2.3.0.0.2.0 uinteger 1
.1.2.3.0.0.2.1 uinteger 2
.1.2.3.0.0.3.0 octet_str "0.123"
.1.2.3.0.0.3.1 octet_str "0.456"
.1.2.3.0.0.3.2 octet_str "9.999"
.1.2.3.0.1 octet_str "baz"
.1.2.3.0.2 integer 234
```

Config:
```
[[inputs.snmp2]]
	uris = [ "snmp2c://localhost:1161" ]

	name = "normalvalues"
	[[inputs.snmp2.field]]
		name = "hostname"
		is_tag = true
		oid = ".1.2.3.0.1"
	[[inputs.snmp2.field]]
		name = "loadavg"
		oid = ".1.2.3.0.2"
		conversion = "float(2)" # converts 123 to 1.23

	[[inputs.snmp2.table]]
		name = "mytable"
		inherit_tags = ["hostname"]
		[[inputs.snmp2.table.field]]
			name = "tablefield1"
			oid = ".1.2.3.0.0.1"
			is_tag = true
		[[inputs.snmp2.table.field]]
			name = "tablefield2"
			oid = ".1.2.3.0.0.2"
		[[inputs.snmp2.table.field]]
			name = "tablefield3"
			oid = ".1.2.3.0.0.3"
			conversion = "float" # converts "0.123" to 0.123
```

Output:
```
* Plugin: snmp2, Collection 1
> normalvalues,agent_host=localhost,host=myhost,hostname=baz loadavg=2.34 1466400920000000000
> mytable,agent_host=localhost,host=myhost,hostname=baz,tablefield1=bar tablefield2=2i,tablefield3=0.456 1466400920000000000
> mytable,agent_host=localhost,host=myhost,hostname=baz,tablefield1=foo tablefield2=1i,tablefield3=0.123 1466400920000000000
```

### Direction
The PR is incomplete. I wasn't sure if this would be accepted at all, and if it was, should it replace the existing SNMP plugin, or go in as an alternate plugin.
Depending on the responses, I can clean up the code a little bit, add more documentation, more tests, and take care of the other stuff like changelog & readme.

### Required for all PRs:

- [ ] CHANGELOG.md updated
- [X] Sign [CLA](https://influxdata.com/community/cla/) (if not already signed)
- [ ] README.md updated (if adding a new plugin)

sparrc commented:
This looks great, thank you very much @phemmer, I think this will be a good improvement.

leave it named snmp2 for now and I'll decide how best to go about maintaining the both of them.

Does this plugin also support snmp v3? v1? if not, could it?

Please also add some documentation of each of the arguments, thanks!

sparrc commented on plugins/inputs/snmp2/snmp.go:
```diff
@@ -0,0 +1,404 @@
+
+const description = `Retrieves SNMP values from remote agents`
+const sampleConfig = `
+[[inputs.snmp2]]
```
[[inputs.snmp2]] doesn't need to be here, it gets added automatically

sparrc commented on plugins/inputs/snmp2/snmp.go:
```diff
@@ -0,0 +1,404 @@
+const description = `Retrieves SNMP values from remote agents`
+const sampleConfig = `
+[[inputs.snmp2]]
+	uris = [ "snmp2c://public@127.0.0.1:1161" ]
```
this section should be indented by 2 spaces, not tabs.

phemmer commented:
Updated to address comments. Still WIP though (more cleanup, more tests, & documentation)

> Does this plugin also support snmp v3? v1? if not, could it?

It didn't, but does now.

Are there any objections to having the tests start up a `snmpd` process if it's available on the host machine? The tests would be a lot more thorough if they could speak to a real snmp daemon.
I'd still have it run as many tests as possible if `snmpd` is not available, and it'd just skip the ones that do require it.

sparrc commented:
@titilambert do you mind reviewing this as well?

sparrc commented:
@phemmer I don't mind starting an snmpd process

titilambert commented:
@sparrc As you guess, unfortunately, I'm really really busy for a few months (and I will for one or two more months :/ )
Of course, this plugin is simpler (it's hard to make more complex than the old one :D).
But, I think It's better to have only one plugin for one thing. Maybe it's better to try to merge both (I don't know how for now).

I will try a quick review today.
But quickly: about unit test, why @phemmer do you ship binaries inside git repo ? circleCI permits you to use docker container. Also, you can use snmpsim (http://snmpsim.sourceforge.net/) which is really cool for unit tests

sparrc commented:
snmpd will not be within the repo, but it could be used if it's available

phemmer commented:
> why @phemmer do you ship binaries

Because whoops :-)

titilambert commented:
@phemmer I just don't get why you rewrite something new. Why not patch the first one ?
Maybe the best think to do is to open a new discussion (github issue) and list all use cases.
Then try to define a config structure to try get some simpler and powerful.
And make it !
@phemmer @sparrc what do you think about ?

sparrc commented:
@titilambert there are quite a few issues and feature requests for the current SNMP plugin. The current plugin is rather complicated and that makes it difficult for me to maintain or have any hope of adding features. If it's difficult for me, then I can only imagine the difficulties for someone in the open-source community.

phemmer commented:
@titilambert largely because of what @sparrc mentioned. Changing the config syntax requires changing the structs, and once you change those, you have to change the code that uses them, and the simpler syntax I proposed was radically different from the existing syntax, which means radically different code. Sometimes it's easier to just start from scratch (and less buggier in the end).

We use SNMP very heavily where I work, and because the existing SNMP plugin doesn't work for us, we needed write a new one anyway. The plan was to propose this upstream, and if if gets merged then great, but otherwise we'd just use the plugin internally.

rgelobterFH commented:
I would agree that while the current SNMP plugin is great and works, it has limitations and can be very confusing to get it initially configured. I think its for that reason a lot of people are still using influxsnmp with grafana/influxdb instead of telegraf. One thing that is nice about influxsnmp is that it was understood and meant for network devices. It knew you wanted to collect ifHCInOctets, ifHCOutOctets, etc and you could specify specific interfaces if you wanted to vs having to collect on every interface regardless. 

I'd personally like to see more flexibility long term in the SNMP plugin.

sunnos9 commented:
Hello guys, new to using the snmp plugin for telegraf.  I need to query snmp tables and preferably add them as tables into InfluxDB, I came across this discussion, what is the suggested path now?

phemmer commented:
I still need to do cleanup & docs. Probably won't get to it until Wednesday at the earliest. Worst case, this weekend as it's a nice 3 day weekend, so plenty of time.

sunnos9 commented:
thank you @phemmer , will this go upstream as snmp or snmp2, will this be part of the next telegraf stable build?

sparrc commented:
@sunnos9 that's yet to be determined, I wouldn't expect this in a stable build for at least a month.

kaey commented:
I've started testing this plugin on our network and here are a few things:

1. Default value of 50 in gs.MaxRepetitions (https://godoc.org/github.com/soniah/gosnmp#GoSNMP) causes snmp response packet to exceed 1500 mtu on some oids. This revealed a bug in our network config, but I suspect that people would want to reduce packet fragmentation, so I'd suggest to make MaxRepetitions field configurable.

2. Storing string value in field, rather than tag is currently done incorrectly - since value is of type []byte it doesn't get wrapped in quotes, which results in error from influxdb. Fix is trivial:
```
 func fieldConvert(conv string, v interface{}) interface{} {
        if conv == "" {
+               switch vt := v.(type) {
+               case []byte:
+                       return string(vt)
+               }
                return v
        }
```

3.  https://github.com/influxdata/telegraf/pull/1389/files#diff-e352e9fa6539e0f5dea40acef9c2e8d6R135
You can set all default values on struct initialization, toml will not override them, if they are not present in config file, no need to do those -1 hacks. This will allow you to remove some ifs from getConnection.


I will need to work with ~5k hosts, so I'll try to make this plugin query agents concurrently and will report back after broader testing.

phemmer commented:
Did a bunch of code cleanup (including support for max repetitions, and byteslice fields), and added some more tests.

I think this is about ready for formal review.
Though a couple outstanding points for discussion:

1. Plugin name - I don't like the name "snmp2". I think it's likely to be confused to be in some relation with SNMP protocol version 2. Maybe "snmp_ng"?

2. `name` vs `name_override` - The plugin has a `name` parameter, which sets the measurement name. On the top level fields, this isn't needed as `name_override` does the same thing. However on tables it is needed as `name_override` has no effect on them. So I left `name` on the top level for consistency. What do we want to do here?
  * Leave as is
  * Remove `name` on the top level fields
  * Remove `name` on the top level fields & rename `name` to `name_override` on the tables?

kaey commented:
Thanks for fixes. Another issues I've encountered: 
1. After tonight's cron jobs plugin started emitting error "Out of order response" for all hosts. Restarting telegraf returned everything back to normal.
You need to close connection gs if gatherTable returned error.
2. From error log it was unclear which host failed. My suggestion is to prefix errors from getConnection and gatherTable with agent url and change errors join string from ", " to "\n".

phemmer commented:
> After tonight's cron jobs plugin started emitting error "Out of order response" for all hosts. Restarting telegraf returned everything back to normal.

Likely like some sort of bug in the `soniah/gosnmp` library (that's who's generating the error). Unfortunately, without sticking a ton of debugging code in the library, the only thing I can think of to figure it out is a packet capture. Is that something you'd be able to do? Get a packet capture of a few of the requests & responses?

> From error log it was unclear which host failed. My suggestion is to prefix errors from getConnection and gatherTable with agent url and change errors join string from ", " to "\n".

Agree with the prefix thing. However this does raise a good point. I think what I'm going to do is just print the errors as they are received instead of collecting them and dumping them all to the telegraf core to print out. Plugins with partial success/failure is unfortunately something telegraf doesn't handle very well.

kaey commented:
> Likely like some sort of bug in the soniah/gosnmp library.
I understand, problem can be even lower - in OS network layer. That's why I suggested closing socket on error.

Patch looks like this
```
diff --git a/plugins/inputs/snmp2/snmp.go b/plugins/inputs/snmp2/snmp.go
index bfe16c7..ac717e6 100644
--- a/plugins/inputs/snmp2/snmp.go
+++ b/plugins/inputs/snmp2/snmp.go
@@ -217,10 +217,13 @@ func (s *Snmp2) Gather(acc telegraf.Accumulator) error {
 		}
 		if err := s.gatherTable(acc, gs, t, false); err != nil {
 			errs = append(errs, err)
+			s.closeConnection(agent)
 		}
 		for _, t := range s.Tables {
 			if err := s.gatherTable(acc, gs, t, true); err != nil {
 				errs = append(errs, err)
+				s.closeConnection(agent)
+				break
 			}
 		}
 	}
@@ -340,6 +343,7 @@ type snmpConnection interface {
 	//BulkWalkAll(string) ([]gosnmp.SnmpPDU, error)
 	Walk(string, gosnmp.WalkFunc) error
 	Get(oids []string) (*gosnmp.SnmpPacket, error)
+	Close() error
 }
 
 // gosnmpWrapper wraps a *gosnmp.GoSNMP object so we can use it as a snmpConnection.
@@ -347,9 +351,14 @@ type gosnmpWrapper struct {
 	*gosnmp.GoSNMP
 }
 
+func (gsw gosnmpWrapper) Close() error {
+	return gsw.Conn.Close()
+}
+
 func (gsw gosnmpWrapper) Host() string {
 	return gsw.Target
 }
+
 func (gsw gosnmpWrapper) Walk(oid string, fn gosnmp.WalkFunc) error {
 	if gsw.Version == gosnmp.Version1 {
 		return gsw.GoSNMP.Walk(oid, fn)
@@ -358,6 +367,15 @@ func (gsw gosnmpWrapper) Walk(oid string, fn gosnmp.WalkFunc) error {
 	}
 }
 
+func (s *Snmp2) closeConnection(agent string) error {
+	if gs, ok := s.connectionCache[agent]; ok {
+		delete(s.connectionCache, agent)
+		return gs.Close()
+	}
+
+	return nil
+}
+
 // getConnection creates a snmpConnection (*gosnmp.GoSNMP) object and caches the
 // result using `agent` as the cache key.
 func (s *Snmp2) getConnection(agent string) (snmpConnection, error) {
```

I've captured a dump, but didn't find any anomalies, I can send it to you privately if you want to take a look.

>I think what I'm going to do is just print the errors as they are received
SGTM

phemmer commented:
> I've captured a dump, but didn't find any anomalies, I can send it to you privately if you want to take a look.

That'd be great if you could. patrick.hemmer@gmail

kirillkovalenko commented:
Is there ETA for this plugin?

phemmer commented:
@kaey found the cause of the `Out of order response` errors. It is indeed a bug in the gosnmp library (raised soniah/gosnmp#68 for it).

I'll work on a patch for the issue and submit it upstream (in the next day or two, not tonight).
I'll also put in some code to get a new connection.

Also after looking at the use case you sent me, I think there's another feature that would be highly beneficial to include. A `inherit_tags = [...]` option on the table config, which will inherit the specified top-scope tags. This is useful for example if you want to fetch the remote agent's host name, and use that as a tag in all the table measurements.

Also due to the multiple errors per interval thing, I think I'm going to make this PR wait for the resolution to #1446 (which I am also working on).

Gelob commented:
@kirillkovalenko at the least a month out before appearing in a stable build. https://github.com/influxdata/telegraf/pull/1389#issuecomment-228905667

lizaoreo commented:
I've got my dashboard buildout on hold until then :)  Excited for this new SNMP plugin.

phemmer commented:
Ok, I've pushed an update which reconnects on any error from gosnmp, and includes the `inherit_tags` option.

I've also submitted a PR on gosnmp to fix the underlying issue. Until that PR is merged, I would recommend setting `Retries = 0`. Otherwise you can end with a lot of network spam until enough dups are encountered to raise an error reset the connection.

sparrc commented:
@phemmer is this ready to go from your perspective? Could you writeup a README as well?

phemmer commented:
Kinda. I was going to wait for #1446 to address some logging issues. But I could somewhat address them without that, and then fix it properly in the PR for #1446.

There are 2 outstanding questions though:
1. Naming. Do we want to call it `snmp2`, or something like `snmp_ng` instead? I just worry that `snmp2` will confuse it with SNMP protocol version 2.
2. The `name` parameter. The plugin has a `name` parameter which controls the metric name. This kinda duplicates the `name_override` parameter, but I think is necessary at least for tables because otherwise all the measurements (all the tables) would have the same name (maybe that's how it should be?). Do we want to leave it as is? Remove the `name` parameter from the top level (non-table) attributes? Something else?

sparrc commented:
don't worry about the `name` parameter. It's OK if it does the same thing in certain cases as `name_override`. I don't think that all tables should be named the same thing, that should either be settable or dynamic.

phemmer commented:
Ok, updated PR with error tweak so that it indicates which agent was being processed when the error was generated. Will readdress error handling in/after #1446
Also updated README & CHANGELOG. PR should be ready to go.

sparrc commented on plugins/inputs/snmp2/snmp.go:
```diff
@@ -0,0 +1,623 @@
+    oid = ".1.2.3.0.1.201"
+
+  [[inputs.snmp2.table]]
+    name = "remote_servers"
```
the `name` parameter under snmp.tables is a _measurement_ name, correct?

phemmer replied:
Correct.

sparrc commented:
Beyond updating the repo-wide README, you also need to write a README specifically for this plugin now... see step 5 here: https://github.com/influxdata/telegraf/blob/master/CONTRIBUTING.md

sparrc commented:
One other comment @phemmer. Throughout the sampleConfig, you have default values defined, but then you have _different_ default values in the comments.

My recommendation is to remove the `# Default` comments, and just put the default values into the sampleConfig string.

phemmer commented:
> Throughout the sampleConfig, you have default values defined, but then you have different default values in the comments.

Doing it that way make it confusing which values in the sample are shown as default, and which are shown as example usage. I was trying to focus on example usage rather than showing defaults. But I have adjusted it.

I have also added a more README on the plugin with more detailed documentation.

Infiniverse commented:
Cool cool cool. I didn't know that this existed - it's in out backlog to test and investigate.

sparrc commented on plugins/inputs/snmp2/README.md:
```diff
@@ -0,0 +1,130 @@
+  [[inputs.snmp2.field]]
+    name = "loadavg"
+    oid = ".1.2.3.0.3"
+  	conversion = "float(2)"
```
this should be indented by spaces, no tabs

sparrc commented on plugins/inputs/snmp2/snmp.go:
```diff
@@ -0,0 +1,624 @@
+				return nil, Errorf(err, "performing get")
+			} else if pkt != nil && len(pkt.Variables) > 0 && pkt.Variables[0].Type != gosnmp.NoSuchObject {
+				ent := pkt.Variables[0]
+				ifv[ent.Name[len(oid):]] = fieldConvert(f.Conversion, ent.Value)
```
can you explain what `ifv[ent.Name[len(oid):]]` is doing? why not just `ifv[ent.Name]`?

phemmer replied:
The `ifv` variable is a map of the values in each row for a single column. The map's key is the field index. The field index is the portion of the OID after the table prefix.

Lets say you have the following data:
```
.1.1.1.2.0: "aaa"
.1.1.1.2.1: "bbb"
.1.1.1.3.0: "xxx"
.1.1.1.3.1: "yyy"
```
Your fields are `foo=.1.1.1.2` and `bar=.1.1.1.3`.
Our end result that we want to achieve is:
```
foo="aaa",bar="xxx"
foo="bbb",bar="yyy"
````

So in the above, your table is `.1.1.1`.
One column is `.1.1.1.2` with row values `.1.1.1.2.0=aaa` and `.1.1.1.2.1=bbb`. The field indexes are `.0` and `.1`, and `ifv` becomes `map{".0":"aaa", ".1":"bbb"}`.
The next column is `.1.1.1.3`, with row values `.1.1.1.3.0=xxx` and `.1.1.1.3.1=yyy`, and `ifv` becomes `map{".0":"xxx", ".1":"yyy"}`.
The result is that you have a bunch of `ifv` variables, one for each column, but they all share the same keys (which allows you to reorganize the data into rows instead of columns).

```
map{".0":"aaa", ".1":"bbb"} // prefix=".1.1.1.2"
map{".0":"xxx", ".1":"yyy"} // prefix=".1.1.1.3"
```
But if you just did `ifv[ent.Name]`, you'd end up with
```
map{".1.1.1.2.0":"aaa", ".1.1.1.2.1":"bbb"}
map{".1.1.1.3.0":"xxx", ".1.1.1.3.1":"yyy"}
```
Which is a lot harder to turn into our end result:
```
map{".1.1.1.2":"aaa", ".1.1.1.3":"xxx"} // index=".0"
map{".1.1.1.2":"bbb", ".1.1.1.3":"yyy"} // index=".1"
```

And no, you can't just use the last digit in the OID as the index. What if you had the following data instead:
```
.1.1.1.2.4.5.6: "aaa"
.1.1.1.2.7: "bbb"
.1.1.1.3.4.5.6: "xxx"
.1.1.1.3.7: "yyy"
```
*(This is a common method of embedding strings in OIDs, known as a "string index". For example `.104.101.108.108.111` is the string `hello`)*

Your table is still `.1.1.1`, and your fields are still `foo=.1.1.1.2` and `bar=.1.1.1.3`. If you did `ifv[ent.Name]`, you'd end up with:
```
map{".1.1.1.2.4.5.6":"aaa", ".1.1.1.2.7":"bbb"}
map{".1.1.1.3.4.5.6":"xxx", ".1.1.1.3.7":"yyy"}
```
How would you then convert that into:
```
map{".1.1.1.2":"aaa", ".1.1.1.3":"xxx"} // index=".4.5.6"
map{".1.1.1.2":"bbb", ".1.1.1.3":"yyy"} // index=".7"
```

---

Now all this is really only for table data. On this specific line of the code, we aren't handling table data, we're handling the fields that were configured at the top level of the plugin. But lets say you have the configuration:
```
[[inputs.snmp2]]
    [[inputs.snmp2.field]]
        name = "f1"
        oid = ".2.2.2.0"
    [[inputs.snmp2.field]]
        name = "f2"
        oid = ".2.2.3.4.5"
    [[inputs.snmp2.field]]
        name = "f3"
        oid = ".2.2.6"
```
We don't want to output this to influxdb as
```
snmp2 f1="foo"
snmp2 f2="bar"
snmp2 f3="baz"
```
We want to output it as
```
snmp2 f1="foo",f2="bar",f3="baz"
```

We accomplish this by pretending that the index on all three of those fields is the same (specifically the empty string `""`, but anything will do as long as it's the same), and so we end up with 3 maps:
```
map{"":"foo"} // prefix="2.2.2.0"
map{"":"bar"} // prefix="2.2.2.3.4.5"
map{"":"baz"} // prefix="2.2.6"
```
And when we flip this so that we have maps of row values instead of maps of column values, we end up with:
```
map{"2.2.2.0":"foo", "2.2.2.3.4.5":"bar", "2.2.6":"baz"}
```
Doing it this way is not only to output a single row to influxdb, but it also lets us share code between non-table and table data.

sparrc commented on plugins/inputs/snmp2/snmp.go:
```diff
@@ -0,0 +1,624 @@
+				if len(ent.Name) <= len(oid) || ent.Name[:len(oid)+1] != oid+"." {
+					return NestedError{} // break the walk
+				}
+				ifv[ent.Name[len(oid):]] = fieldConvert(f.Conversion, ent.Value)
```
same as above comment

phemmer replied:
same answer as above :-)

sparrc commented:
the main feature I'd still like to see is support for MIB names. In my mind it would look something like this:

Support  looking up a mib name using snmptranslate: `snmptranslate -On -IR ifOutOctets.1`
```
[[inputs.snmp2.field]]
  mib = "ifOutOctets.1"
```

If given an OID without a name, use snmptranslate to get the MIB name: `snmptranslate -Os .1.3.6.1.2.1.2.2.1.16.1`
```
[[inputs.snmp2.field]]
  oid = ".1.3.6.1.2.1.2.2.1.16.1"
```

obviously, the user can still add their own `name` and this would override the actual MIB name.

sparrc commented:
secondly, I am curious if it's possible to add support for querying an snmptable directly, rather than having to enumerate each field that the user wants.

For example, something like this:

```
  [[inputs.snmp2.table]]
    name = "mytable"
    inherit_tags = ["hostname"]
    table = "ifXTable"
```

Telegraf would then walk the ifXTable and gather metrics therein.

sparrc commented on plugins/inputs/snmp2/snmp.go:
```diff
@@ -0,0 +1,624 @@
+	rows := map[string]RTableRow{}
+
+	tagCount := 0
+	for _, f := range t.Fields {
```
here I think we can check that the OID is a valid table field with something like:

```
if walk && strings.HasSuffix(f.Oid, ".0") {
    // log error and continue?
}
```

schotlandzegtja commented:
I am going to mention this here since this appears to be the PR for the impending SNMP plugin rewrite.

In using Telegraf, I encountered a few devices returning a 'tooBig' error; goSNMP doesn't seem to handle this well or at all and it doesn't provide a mechanism for altering the limit to the number of OIDs sent at any one time. The limit is a hard limit of 60.

I raised an issue and submitted a PR to change this. The maintainer appears to be doing some more work on top of my PR but the change will likely be adopted today.

I will likely raise a PR to allow manual per-host configuration of the 'MaxOids' at some point if no one else wants to work it in, but I am not sure if I should wait until this PR is completed.

https://github.com/soniah/gosnmp/issues/72 
https://github.com/soniah/gosnmp/pull/73

Gelob commented:
@phemmer I'm looking at the updated README in trying to test this out. Can you give a brief config example of how I would handle multiple devices (agents) with different SNMP communities? Would I just have duplicate [[inputs.snmp2]] sections?

sparrc commented:
@Gelob you can define multiple [[inputs.snmp2]] inputs. Every plugin has an `interval` option available, which is that same as that in the overall agent config, but just needs to be defined underneath the input.

ie,
```
[[inputs.snmp]]
  interval = "60s"
  ...
```

phemmer commented:
@schotlandzegtja This plugin is unaffected by the issue you describe. The original snmp plugin would pass multiple OIDs to a single SNMP get call. This plugin heavily utilizes bulk walk instead of gets. There is some get, but it's usage is minimal, and it only gets a single OID at a time.

There are a few performance optimizations that the plugin can benefit from, a goroutine for each agent is one big one, but multiple OIDs in a single get call would be another. If/when that is implemented, yes, we'll need to make sure we properly handle the issue. Such as a backoff on the number OIDs in one get call until we no longer get the tooBig error. We can then cache the value and reuse it.

phemmer commented:
@sparrc 
> the main feature I'd still like to see is support for MIB names

I'm in favor of this. However I would argue against separate config params for OID number vs name. We should be able to simply look at the value and convert it if necessary.

phemmer commented:
@sparrc 

> secondly, I am curious if it's possible to add support for querying an snmptable directly, rather than having to enumerate each field that the user wants.

Possible yes. I'm not opposed to it as long as we don't drop support of manually specifying fields. 2 reasons why.
1. The user may not want all fields of the table. The table could be huge, and result in a lot of unwated data getting shoved into influxdb (or whatever backend).
2. The user may have data which spreads multiple tables. For example, they may have 2 tables which track network interface statistics. One table reports metrics such as layer 1 counters (bytes in, bytes out, line errors, etc). Another table might report layer 2+ counters (packets, active tcp sessions, etc). But both tables are indexed by the interface ID. Allowing manually specifying fields lets output specific fields from each table as if they were the same table.

sparrc commented:
> I'm in favor of this. However I would argue against separate config params for OID number vs name. We should be able to simply look at the value and convert it if necessary.

even better

> Possible yes. I'm not opposed to it as long as we don't drop support of manually specifying fields. 2 reasons why.

:+1:

Gelob commented:
@phemmer I'm having an issue where gosnmp is literally returning \<nil\> for my sysName OID pull. Is there something I'm missing? Seems like its expecting an integer for some reason although my pull of ifAlias works. If you want a packet capture let me know.

telegraf.conf
```
[[inputs.snmp2]]
  agents = [ "1.1.1.1" ]
  version = 2
  community = "blah"

  name = "bwStats"
  [[inputs.snmp2.field]]
    name = "devicename"
    oid = ".1.3.6.1.2.1.1.5"
    is_tag = true

  [[inputs.snmp2.table]]
    name = "bwStats"
    inherit_tags = [ "devicename" ]
    [[inputs.snmp2.table.field]]
      name = "ifAlias"
      oid = ".1.3.6.1.2.1.31.1.1.1.18"
      is_tag = true
        [[inputs.snmp2.table.field]]
        name = "ifHCInOctets"
        oid = ".1.3.6.1.2.1.31.1.1.1.6"

        [[inputs.snmp2.table.field]]
        name = "ifHCOutOctets"
        oid = ".1.3.6.1.2.1.31.1.1.1.10"

        [[inputs.snmp2.table.field]]
        name = "ifHCInUcastPkts"
        oid = ".1.3.6.1.2.1.31.1.1.1.7"

        [[inputs.snmp2.table.field]]
        name = "ifHCOutUcastPkts"
        oid = ".1.3.6.1.2.1.31.1.1.1.11"

        [[inputs.snmp2.table.field]]
        name = "ifInErrors"
        oid = ".1.3.6.1.2.1.2.2.1.14"

        [[inputs.snmp2.table.field]]
        name = "ifOutErrors"
        oid = ".1.3.6.1.2.1.2.2.1.20"
```

Output from telegraf -config telegraf.conf -test

`bwStats,agent_host=1.1.1.1,devicename=<nil>,ifAlias=testinterface,ifHCInOctets=777078506i,ifHCInUcastPkts=978081i,ifHCOutOctets=549342425i,ifHCOutUcastPkts=644632i,ifInErrors=0i,ifOutErrors=0i 1470435158000000000`

Partial output from modified gosnmp example2.go to pull the same OID. I removed the community portion.
go run name2.go
```
Packet sanity verified, we got all the bytes (52)
parseRawField: version
0       02 01 01 04 10 73 69 6e
8       30 31 62 6e 74 76 7a 6f
Parsed version 1
parseRawField: request id
0       02 04 5a 62 67 73 02 01
8       00 02 01 00 30 0d 30 0b
requestID: 1516398451
parseRawField: error-status
0       02 01 00 02 01 00 30 0d
8       30 0b 06 07 2b 06 01 02
errorStatus: 0
parseRawField: error index
0       02 01 00 30 0d 30 0b 06
8       07 2b 06 01 02 01 01 05
error-index: 0

=== unmarshalVBL()
0       30 0d 30 0b 06 07 2b 06
8       30 0d 30 0b 06 07 2b
vblLength: 15

STARTING a varbind. Cursor 2
0       30 0b 06 07 2b 06 01 02
8       30 0b 06 07 2b
parseRawField: OID
0       06 07 2b 06 01 02 01 01
8       06 07 2b
OID: .1.3.6.1.2.1.1.5
decodeValue: value
0       81 00
decodeValue: type is NoSuchInstance
decodeValue: value is <nil>
0: oid: .1.3.6.1.2.1.1.5 number: 0

```

snmpwalk output (same output for version1 too)
```
snmpwalk -v2c -c blah 1.1.1.1 1.3.6.1.2.1.1.5
SNMPv2-MIB::sysName.0 = STRING: blah01-edge01
```

kaey commented:
If you look closely at snmpwalk output you'll notice it's sysName.0.
Thus OID should be 1.3.6.1.2.1.1.5.0

Gelob commented:
@kaey thanks, I've never felt more dumb for commenting but that was exactly it.

sparrc commented:
@phemmer would you be able to get in these changes by next Thursday, the 18th? We'll be cutting a 1.0 release candidate shortly after that so that will be the deadline for getting the new SNMP plugin in for 1.0.

Let me know if there's anything I can do to help.

phemmer commented:
The oid number/name conversion would be easy to do sure.
The not needing to specify fields of a table is a little hairy. I don't think implementation will be hard, but design might be (not making the config a mess). Will take a poke at it this weekend.

phemmer commented:
I think I've got a decent solution to this.
In the configuration, an `oid` parameter can be specified on a table. If provided, we extract all the table columns from the MIB and automatically add them as if you had added fields to the config. Meaning we don't do a walk to discover the fields, they have to be defined in the MIB.

I've coded it up, and will work on polishing it and writing tests in the next few days.

sparrc commented:
sounds good, thanks @phemmer, once I merge https://github.com/influxdata/telegraf/pull/1626, you can rebase off that and rename your plugin to `snmp`

phemmer commented:
The docs still need more work (will finish them tomorrow), but wanted to publish as the code & tests should be complete, so it should be functional if you want to try it and provide feedback.

sparrc commented:
there appears to be a data race in the code (or possibly just in the test code)

sparrc commented on plugins/inputs/snmp2/snmp.go:
```diff
@@ -0,0 +1,788 @@
+	// The SNMP agent to query. Format is ADDR[:PORT] (e.g. 1.2.3.4:161).
+	Agents []string
+	// Timeout to wait for a response. Value is anything accepted by time.ParseDuration().
+	Timeout string
```
the Timeout parameter should be in the output of SampleConfig. Also use an `internal.Duration` type, which calls time.ParseDuration at the time that the toml is parsed.

phemmer replied:
Done

sparrc commented on plugins/inputs/snmp2/snmp.go:
```diff
@@ -0,0 +1,788 @@
+			return nil, Errorf(err, "parsing timeout")
+		}
+	} else {
+		gs.Timeout = time.Second * 1
```
This default seems to be a bit on the short side to me, especially when we're talking about SNMP and potentially very old and/or slow devices. Maybe default to 5s?

phemmer replied:
Done

phemmer commented:
> there appears to be a data race in the code (or possibly just in the test code)

Was just the test code. But I fixed it.

sparrc commented:
looks good and is working as expected from my perspective. Thanks a bunch @phemmer, I'm going to get this merged so I can get it into the 1.0 release candidate branch

schotlandzegtja commented:
From the user community, I thank you all for your effort!

3fr61n commented:
Nice plugin, is working in the first attempt :) 

Question, is there any way to add a filter in a table?

I'm trying to get interface statistics from a router, but I'm only interested in a few of them, so I would like to have kind of regex or filter that limit which interfaces should query

Perhaps something like


 [[inputs.snmp.table]]
    name = "interfaces"
    inherit_tags = [ "hostname" ]
    [[inputs.snmp.table.field]]
      name = "ifDescr"
      oid = ".1.3.6.1.2.1.2.2.1.2"
      filter_regex = "ge-.*"
      is_tag = true
    [[inputs.snmp.table.field]]
      name = "ifInOctets"
      oid = ".1.3.6.1.2.1.2.2.1.10"
    [[inputs.snmp.table.field]]
      name = "ifOutOctets"
      oid = ".1.3.6.1.2.1.2.2.1.16"

Regards

phemmer commented:
@3fr61n Yes, see the `tagdrop` config param: https://github.com/influxdata/telegraf/blob/master/docs/CONFIGURATION.md

3fr61n commented:
thanks @phemmer  it works

and is there any way (if not today, even in the future), to avoid gathering the whole table and just collect data for the 'filtered' items?

I'm asking this because in our case gathering the whole table (1000s of interfaces, when we just want to monitor a couple of hundred), could overload the router cpu and possibly have service impact

Thanks in advance

phemmer commented:
@3fr61n Not likely. Doing what you are requesting would require performing SNMP GET calls. The plugin does SNMP bulk queries, as they're far more efficient. Switching to SNMP GET just to add this filtering capability would likely increase the load on the router, not decrease it. Processing a small amount of bulk requests will be less intensive than thousands of get requests.

Also SNMP will never cause performance traffic performance degradation on a router. See https://www.cisco.com/c/en/us/support/docs/ip/simple-network-management-protocol-snmp/7270-ipsnmphighcpu.html

> SNMP is a low priority process as far as the CPU scheduler is concerned, so another process requiring CPU resources takes priority. Therefore, while CPU spikes occur in this scenario, they should not affect performance.

3fr61n commented:
I understand your point of view, however in my opinion depending of the scenario one approach is more efficient than the other,  if you are interested in the majority of the entries of the table, a snmpbulk is the best option, and in the other case a snmp get fits better

Regarding the cisco link, it's a little tricky, because if SNMP polling is really harmless why they recommend (below in that link) about how to reduce the data to be poll using views, or how to configure the server to specify oids  ;)

The problem behind the router is not only the snmp agent, sometimes the data could be store/handled by other processes inside the router, so a inter process interaction begins, and depending of the amount of interaction cpu spikes could arise


BTW:  Thanks for you prompt answer 👍

kaey commented:
> Also SNMP will never cause performance traffic performance degradation on a router.

This is incorrect. While there is separation between routing engine (RE) and packet forwarding engine (PFE), daemons, such as bgpd and snmpd, run on a single RE, thus noise from one will affect the other.

> and in the other case a snmp get fits better

snmpget requires roundtrip per every single metric. I don't think there's a situation, where get is more efficient then bulkwalk, unless you query 5 metrics vs 500.

phemmer commented:
> This is incorrect. While there is separation between routing engine (RE) and packet forwarding engine (PFE), daemons, such as bgpd, snmpd and sshd, run on a single RE, thus noise from one will affect the other.

Correct, I was referring to routing traffic. The other services *might* see the impact from snmp usage, but the snmp usage will be very brief (and a few thousand interfaces is very unlikely to cause any noticeable spike in cpu), and thus have no effective impact.

3fr61n commented:
@kaey  In a test doing continuous polling to any table inside jnxBgpM2Peer, when you have more than 4k sessions, you'll see constant high cpu usage from routing daemons, and if you push further  you can even affect the sessions.

Regarding 

> snmpget requires roundtrip per every single metric. I don't think there's a situation, where get is more efficient then bulkwalk, unless you query 5 metrics vs 500.

I'm totally agree, and In the case of interfaces for service providers this is basically what happens... that you have many customers/services on subinterfaces, but you are only interested on the physical interface statistic.

sparrc commented:
@3fr61n So does an SNMP get solve your use-case? I believe using the `snmp.field` config does a simple SNMP get. @phemmer correct me if I'm wrong.

phemmer commented:
@sparrc `snmp.field` only does a get if it's in the top level config. I.E. we're fetching a single value. All table metrics are fetched using bulk.

3fr61n commented:
@sparrc  you are correct, however  the counterpart of using snmp.field (besides what @sparrc mentioned)  is that you need to previously know the table index in order to build the full oid,  this approach does not scale because the index could change from time to time.

StianOvrevage commented:
GREAT work on another SNMP plugin! The open-source world is sorely lacking anything that is both configurable and scalable and performant.

Regarding the discussion about snmpwalk and snmpget.

> snmpget requires roundtrip per every single metric. I don't think there's a situation, where get is more efficient then bulkwalk, unless you query 5 metrics vs 500.

That is INCORRECT. Let me prove it:

By issuing `snmpget -m ALL -v2c -c COMMUNITY HOST IF-MIB::ifInMulticastPkts.1 IF-MIB::ifOutMulticastPkts.1 IF-MIB::ifInBroadcastPkts.1 IF-MIB::ifOutBroadcastPkts.1 IF-MIB::ifHCInOctets.1 IF-MIB::ifHCInUcastPkts.1 IF-MIB::ifHCOutOctets.1`

I get all these metrics in one go:

> IF-MIB::ifInMulticastPkts.1 = Counter32: 0
>IF-MIB::ifOutMulticastPkts.1 = Counter32: 0
>IF-MIB::ifInBroadcastPkts.1 = Counter32: 0
>IF-MIB::ifOutBroadcastPkts.1 = Counter32: 0
>IF-MIB::ifHCInOctets.1 = Counter64: 0
>IF-MIB::ifHCInUcastPkts.1 = Counter64: 0
>IF-MIB::ifHCOutOctets.1 = Counter64: 0

AND before you think there is some magic happening with the snmpget tool. This actual tcpdump shows that it does not: 

`18:06:27. IP 10.0.0.4.39562 > host.snmp:  C=community GetRequest(133)  31.1.1.1.2.1 31.1.1.1.4.1 31.1.1.1.3.1 31.1.1.1.5.1 31.1.1.1.6.1 31.1.1.1.7.1 31.1.1.1.10.1
18:06:28 IP host.snmp > 10.0.0.4.39562:  C=community GetResponse(140)  31.1.1.1.2.1=0 31.1.1.1.4.1=0 31.1.1.1.3.1=0 31.1.1.1.5.1=0 31.1.1.1.6.1=0 31.1.1.1.7.1=0 31.1.1.1.10.1=0`

So it is perfectly possible to make snmpgets much more efficient that one-metric-one-request that some of you assume is the default/only way :-D

I did this back a couple of years ago with collectd and observed reduced CPU usage on both server and switch when combining a few known groups of metrics (such as interface metrics which tend to be standard). Wrote about it here: http://www.peritusconsulting.no/articles/2014-06-02-next-generation-monitoring-using-opentsdb.html#Collection (The 2200MHz to 1800MHz on collectd is from combining a few different data type and definitions so that even more metrics are pulled for every SNMP request).

3fr61n commented:
Good catch :)

wlcx commented:
I'm getting errors collecting data from an APC UPS over SNMPv1 - my config is

```
[[inputs.snmp]]
  agents = ["144.32.101.92"]
  version = 1
  community = "public"
  name = "ups"

  [[inputs.snmp.field]]
    name = "name"
    oid = "1.3.6.1.2.1.1.5"
    is_tag = true
    
  [[inputs.snmp.field]]
    name = "output_load"
    oid = "PowerNet-MIB::upsAdvOutputLoad"
```

and telegraf.log displays errors like:
```
ups,agent_host=144.32.101.92,host=vault,name=\u003cnil\u003e output_load= 1472035210000000000
```

Both OIDs resolve properly and are retrievable via `snmpwalk`

StianOvrevage commented:
@wlcx You might want to hide your public IP in the snippet since I'm guessing you are also using `private` as the read-write community and hence (firewall permitting) anyone might be able to shut down your UPS now remotely :)

wlcx commented:
@StianOvrevage read-only and behind a firewall, but snipped nonetheless :)

3fr61n commented:
@wlcx   Did you try to add a '.' in the begining of your oid?

```
 [[inputs.snmp.field]]
    name = "name"
    oid = "1.3.6.1.2.1.1.5"  <----- ".1.3.6.1.2.1.1.5"
    is_tag = true
```

phemmer commented:
Leading dot should is optional. 
@wlcx can you provide the snmpwalk output? At least the portion that covers this value?

wlcx commented:
sure.
```
snmpwalk -Os -c public -v1 <snipped ip> 1.3.6.1.2.1.1.5
iso.3.6.1.2.1.1.5.0 = STRING: "Derrick"
```
P.S. In case you were wondering, I have no idea why the UPS is called Derrick :p

phemmer commented:
Ah, you need to use `1.3.6.1.2.1.1.5.0` in the config, not `1.3.6.1.2.1.1.5`.
I'm guessing the same for your `output_load` field, `PowerNet-MIB::upsAdvOutputLoad.0`.

I'll look into logging a more useful message, such as `no such object at OID $foo`

wlcx commented:
Ah, spot on, that works perfectly. I wonder if it would be worth defaulting to an instance of 0 if one is not given - this must be what snmpwalk etc do.

Thanks all for the help!

phemmer commented:
Comparing to `snmpwalk` isn't a fair assessment as that utility does not do the same thing. The purpose of `snmpwalk` is to retrieve multiple values, we only want 1 value. If you use `snmpget` instead you'll see the same behavior.

lizaoreo commented:
I've been excitedly waiting for this to merge.  Now that it's in, I've got a dumb, somewhat off topic question.  How can I get this on telegraf now without waiting for the official 1.0 release?

phemmer commented:
@lizaoreo It should be in the nightlies (i haven't confirmed though): https://influxdata.com/downloads/

sparrc commented:
We have 1.0 release candidates available which have this new SNMP plugin, it is available here: https://github.com/influxdata/telegraf#installation

Gelob commented:
@3fr61n can you give your config example where tagdrop is working for you? Thanks!

3fr61n commented:
@Gelob 

Here is an example where I just gather physical interface stats and drop any sub-interface stats

'''
[[inputs.snmp]]
  interval = "300s"
  agents = [ "<x.x.x.x>:161" ]
  version = 2
  community = "<xxxxxx>"

  name = "snmp_data"
  [[inputs.snmp.field]]
    name = "hostname"
    oid = ".1.3.6.1.2.1.1.5.0"
    is_tag = true
  [[inputs.snmp.field]]
    name = "uptime"
    oid = ".1.3.6.1.2.1.1.3.0"

  [[inputs.snmp.table]]
    name = "interfaces"
    inherit_tags = [ "hostname" ]
    [inputs.snmp.tagdrop]
      ifDescr = [ "*.*"]
'''

Gelob commented:
@3fr61n thanks. I'm having a terrible time getting tagdrop to work. Even if i put a wildcard in I still see results in my test output. My goal is to try to get rid of the interfaces with no description (IfAlias below) but I can't even seem to get rid of anything with my tagdrop config in just testing it with the wildcard. Do you see anything odd I have going on?

```
[[inputs.snmp]]
  agents = [ "1.1.1.1", "1.1.1.2"]
  version = 2
  community = ""
  interval = "10s"
  name = "bwStats"
  [[inputs.snmp.field]]
    name = "hostname"
    oid = ".1.3.6.1.2.1.1.5.0"
    is_tag = true
  [[inputs.snmp.table]]
    name = "bwStats"
    inherit_tags = [ "hostname" ]
  [[inputs.snmp.tagdrop]]
   ifAlias = [" "]
    [[inputs.snmp.table.field]]
      name = "ifAlias"
      oid = ".1.3.6.1.2.1.31.1.1.1.18"
      is_tag = true
        [[inputs.snmp.table.field]]
        name = "ifHCInOctets"
        oid = ".1.3.6.1.2.1.31.1.1.1.6"
        [[inputs.snmp.table.field]]
        name = "ifHCOutOctets"
        oid = ".1.3.6.1.2.1.31.1.1.1.10"
        [[inputs.snmp.table.field]]
        name = "ifHCInUcastPkts"
        oid = ".1.3.6.1.2.1.31.1.1.1.7"
        [[inputs.snmp.table.field]]
        name = "ifHCOutUcastPkts"
        oid = ".1.3.6.1.2.1.31.1.1.1.11"
        [[inputs.snmp.table.field]]
        name = "ifInErrors"
        oid = ".1.3.6.1.2.1.2.2.1.14"
        [[inputs.snmp.table.field]]
        name = "ifOutErrors"
        oid = ".1.3.6.1.2.1.2.2.1.20"

```
I snipped some of the output lines
```
* Plugin: snmp, Collection 1
* Internal: 10s
> bwStats,agent_host=1.1.1.1,hostname=edge01 ifHCInOctets=0i,ifHCInUcastPkts=0i,ifHCOutOctets=0i,ifHCOutUcastPkts=0i,ifInErrors=0i,ifOutErrors=0i 1473094905000000000
> bwStats,agent_host=1.1.1.1,hostname=edge01,ifAlias=XO\ 10GG\ CIR\ CktID ifHCInOctets=813194263i,ifHCInUcastPkts=14991788i,ifHCOutOctets=6619529230850i,ifHCOutUcastPkts=7054570727i,ifInErrors=0i,ifOutErrors=0i 1473094905000000000
> bwStats,agent_host=1.1.1.1,hostname=edge01 ifHCInOctets=0i,ifHCInUcastPkts=0i,ifHCOutOctets=0i,ifHCOutUcastPkts=0i,ifInErrors=0i,ifOutErrors=0i 1473094905000000000
> bwStats,agent_host=1.1.1.1,hostname=edge01 ifHCInOctets=0i,ifHCInUcastPkts=0i,ifHCOutOctets=0i,ifHCOutUcastPkts=0i,ifInErrors=0i,ifOutErrors=0i 1473094905000000000
> bwStats,agent_host=1.1.1.1,hostname=edge01 ifHCInOctets=0i,ifHCInUcastPkts=0i,ifHCOutOctets=0i,ifHCOutUcastPkts=0i,ifInErrors=0i,ifOutErrors=0i 1473094905000000000
> bwStats,agent_host=1.1.1.1,hostname=edge01 ifHCInOctets=0i,ifHCInUcastPkts=0i,ifHCOutOctets=0i,ifHCOutUcastPkts=0i,ifInErrors=0i,ifOutErrors=0i 1473094905000000000
> bwStats,agent_host=1.1.1.1,hostname=edge01,ifAlias= ifHCInOctets=0i,ifHCInUcastPkts=0i,ifHCOutOctets=0i,ifHCOutUcastPkts=0i,ifInErrors=0i,ifOutErrors=0i 1473094905000000000

```

3fr61n commented:
@Gelob 

I think you have a syntax issue, please check below my config



'''

[[inputs.snmp]]
  interval = "60s"
  agents = [ "172.30.137.212:161" ]
  version = 2
  community = "public"

  name = "snmp_data"
  [[inputs.snmp.field]]
    name = "hostname"
    oid = ".1.3.6.1.2.1.1.5.0"
    is_tag = true
  [[inputs.snmp.field]]
    name = "uptime"
    oid = ".1.3.6.1.2.1.1.3.0"

  [[inputs.snmp.table]]
    name = "interfaces"
    inherit_tags = [ "hostname" ]
    [inputs.snmp.tagdrop]  <--- you have double bracket instead of single bracket
      ifAlias = [ ""]   <-- I think you have a 'blank space'

    [[inputs.snmp.table.field]]
      name = "ifAlias"
      oid = ".1.3.6.1.2.1.31.1.1.1.18"
      is_tag = true


    [[inputs.snmp.table.field]]
      name = "ifHCInOctets"
      oid = ".1.3.6.1.2.1.31.1.1.1.6"
    [[inputs.snmp.table.field]]
      name = "ifHCOutOctets"
      oid = ".1.3.6.1.2.1.31.1.1.1.10"

'''



root@052f081c68de:/etc/telegraf# telegraf -test -config my.conf 
* Plugin: snmp, Collection 1
* Internal: 1m0s
> snmp_data,agent_host=xxxx,host=052f081c68de,hostname=PE2-tf-m120-1-re0 uptime=459894673i 1473096735000000000
> interfaces,agent_host=xxxx,host=052f081c68de,hostname=PE2-tf-m120-1-re0,ifAlias=to\ tf-mx960-2\ ge-1/2/0 ifHCInOctets=587134925i,ifHCOutOctets=628032583i 1473096738000000000
> interfaces,agent_host=xxxx,host=052f081c68de,hostname=PE2-tf-m120-1-re0,ifAlias= ifHCInOctets=173216943i,ifHCOutOctets=195843599i 1473096738000000000
> interfaces,agent_host=xxxx,host=052f081c68de,hostname=PE2-tf-m120-1-re0,ifAlias=to\ ex4600-1\ ge-0/0/12 ifHCInOctets=100416466i,ifHCOutOctets=0i 1473096738000000000
> interfaces,agent_host=xxxx,host=052f081c68de,hostname=PE2-tf-m120-1-re0,ifAlias=to\ tf-m320-1-so-2/2/0 ifHCInOctets=908499298i,ifHCOutOctets=835561372i 1473096738000000000
> interfaces,agent_host=xxxx,host=052f081c68de,hostname=PE2-tf-m120-1-re0,ifAlias=to\ tf-mx960-2\ ge-1/2/1 ifHCInOctets=0i,ifHCOutOctets=0i 1473096738000000000
> interfaces,agent_host=xxxx,host=052f081c68de,hostname=PE2-tf-m120-1-re0,ifAlias=to\ ex4600\ ge-0/0/13 ifHCInOctets=4179898327i,ifHCOutOctets=24018649693i 1473096738000000000
> interfaces,agent_host=xxxx,host=052f081c68de,hostname=PE2-tf-m120-1-re0,ifAlias=to\ tf-mx480-3\ xe-3/3/0 ifHCInOctets=439825082i,ifHCOutOctets=611439616i 1473096738000000000

Gelob commented:
Thanks, that double bracket was it. 👍  Now I just need to get the glob match for the space
